api/services/v1alpha1/gateway_types.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/api/common"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	GatewayServiceName = "gateway"
+	// GatewayInstanceName the name of the Gateway instance singleton.
+	// value should match whats set in the XValidation below
+	GatewayInstanceName = "default-gateway"
+	GatewayKind         = "Gateway"
+)
+
+// Check that the component implements common.PlatformObject.
+var _ common.PlatformObject = (*Gateway)(nil)
+
+// GatewaySpec defines the desired state of Gateway
+type GatewaySpec struct {
+	// Authentication configuration
+	// +optional
+	Auth GatewayAuthSpec `json:"auth"`
+
+	// Certificate management
+	// +optional
+	Certificates GatewayCertSpec `json:"certificates"`
+
+	// Domain configuration for the gateway
+	// +optional
+	Domain string `json:"domain,omitempty"`
+
+	// Namespace where the gateway resources should be deployed
+	// +kubebuilder:default="openshift-ingress"
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// GatewayAuthSpec defines authentication configuration for the gateway
+type GatewayAuthSpec struct {
+	// Authentication mode: "openshift-oauth" | "oidc" | "auto"
+	// +kubebuilder:validation:Enum=openshift-oauth;oidc;auto
+	// +kubebuilder:default="auto"
+	// +optional
+	Mode string `json:"mode,omitempty"`
+
+	// OIDC configuration (required when mode="oidc")
+	// +optional
+	OIDC *OIDCConfig `json:"oidc,omitempty"`
+}
+
+// OIDCConfig defines OIDC provider configuration
+type OIDCConfig struct {
+	// OIDC issuer URL
+	// +kubebuilder:validation:Required
+	IssuerURL string `json:"issuerURL"`
+
+	// Reference to secret containing clientID and clientSecret
+	// +kubebuilder:validation:Required
+	ClientSecretRef corev1.SecretKeySelector `json:"clientSecretRef"`
+}
+
+// GatewayCertSpec defines certificate management configuration
+type GatewayCertSpec struct {
+	// Type of certificate management: "user-provided" | "cert-manager" | "openshift-service-ca"
+	// +kubebuilder:validation:Enum=user-provided;cert-manager;openshift-service-ca
```

Comment:
There are really two types of certificate configurations, one for internal service to service communication and another for ingress.
openshift-service-ca certificates are meant for internal service to service to service communication, so I suggest we not include this option here since we are configuring an ingress certificate for external access.
I would also consider "cert-manager" an implementation detail, would  "self-signed" be more accurate here?
Also need to consider the outcome of this SPIKE: https://issues.redhat.com/browse/RHOAIENG-33685